### PR TITLE
update to solidity 0.8

### PR DESCRIPTION
**Pull request by ralph-pichler**
* update to solidity 0.8
* remove unnecessary `SafeMath` dependency
* remove unnecessary `abicoder` pragma (this should help with source verification)
* replace `ERC20` dependency with `IERC20`

### CODE REVIEW

File: contracts/ERC20SimpleSwap.sol [hunk 1]
```diff
@@ -62,7 +56,7 @@ contract ERC20SimpleSwap {
   );
 
   // the EIP712 domain this contract uses
-  function domain() internal pure returns (EIP712Domain memory) {
+  function domain() internal view returns (EIP712Domain memory) {
     uint256 chainId;
     assembly {
```
Comment (Eknir) [inline contracts/ERC20SimpleSwap.sol:61]: Solidity 0.8 exposes `block.chainid`, see: https://docs.soliditylang.org/en/v0.8.0/units-and-global-variables.html

File: contracts/ERC20SimpleSwap.sol [hunk 2]
```diff
@@ -85,7 +79,7 @@ contract ERC20SimpleSwap {
   }
 
   // recover a signature with the EIP712 signing scheme
-  function recoverEIP712(bytes32 hash, bytes memory sig) internal pure returns (address) {
```
Comment (Eknir) [inline contracts/ERC20SimpleSwap.sol:88]: Could you please explain this change?

**Review by Eknir — COMMENTED** [review]
Thank you for the upgrade. One minor suggestion for improving.
Pay attention to ln 243 => if you want to keep this functioning the same, you should add `unchecked { }`

**Comment by Eknir** [comment]
While you are upgrading, perhaps it makes sense to sanitize https://github.com/ethersphere/swap-swear-and-swindle/blob/master/package.json as well?

File: contracts/ERC20SimpleSwap.sol (hunk 2 above)
Comment (ralph-pichler) [reply contracts/ERC20SimpleSwap.sol:88]: querying the chainid is no longer considered pure in 0.8. as this calls `domain()`,  this function cannot be pure either.

File: contracts/ERC20SimpleSwap.sol (hunk 1 above)
Comment (ralph-pichler) [reply contracts/ERC20SimpleSwap.sol:61]: nice

**Review by Eknir — APPROVED** [review]
LGTM!

**Pull request merged**